### PR TITLE
RFC: Support running npm scripts

### PR DESCRIPTION
**Pull request by arjun024**
[Rendered](https://github.com/paketo-buildpacks/nodejs/blob/rfc-npm-scripts/rfcs/0004-npm-scripts.md)

### CODE REVIEW

File: rfcs/0004-npm-scripts.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+[npm-install process](https://docs.npmjs.com/cli/v6/using-npm/scripts#npm-install).
+
+### Open questions
+
+1. Web framework examples mostly demonstrate cases where a `postinstall` event
+   would be useful. What are some real world cases (if any) where `preinstall`
+   would be used?
+
+1. Do NPM and Yarn need separate env variables or should they read the same set
+   of env vars (e.g. `$BP_NODE_{PRE,POST}INSTALL`)?
```
Comment (kvedurmu) [inline rfcs/0004-npm-scripts.md:130]: I like this configuration a little better. I think users will just look at their `package.json` to figure out what commands/scripts to run pre/post-install, so I think a single set of environment variables is easier.

**Comment by kvedurmu** [comment]
@arjun024 how would layer re-use work with these environment variables?

**Comment by arjun024** [comment]
@kvedurmu This shouldn't have any effect on layer re-use unless the custom script/command changes `node_modules`, `package.json` or `package-lock.json`. For the static files build case for something like angular, it should generate assets to `<app>/dist` and this wouldn't be part of a layer and hence not cached.

File: rfcs/0004-npm-scripts.md (hunk 1 above)
Comment (arjun024) [reply rfcs/0004-npm-scripts.md:130]: You mean the combined one? Which one would you prefer `$BP_NODE_*INSTALL` or `$BP_PACKAGEJSON_*INSTALL`?

**Comment by ryanmoran** [comment]
I worry that this is packing quite a lot of extra functionality into the `{npm,yarn}-install` buildpacks (`npm-install` especially). And more so, I'm not yet convinced that there is evidence of a good use for a pre-install hook.

I'm wondering if we could create just a single new buildpack (something like `node-run-script`) that has a detection requirement on either `npm` or `yarn` and executes the correct command based on what shows up in the build plan. We can consolidate all of this logic into a single buildpack along with singular environment variables to support it.

**Comment by sophiewigmore** [comment]
+1 to the idea of having this logic fall into one new buildpack.
I also have a question: An alternative to using environment variables would be to scan the app for custom events. I'm not that familiar with what these scripts look like. Is it viable to look for scripts in the app for detection, rather than relying on environment variables?

**Comment by arjun024** [comment]
@sophiewigmore See example apps listed in the RFC. It's not quite possible to tell which event is to be executed - differs for different frameworks or in other cases the user defines their own events

**Review by ryanmoran — APPROVED** [review]

**Comment by kvedurmu** [comment]
> @kvedurmu This shouldn't have any effect on layer re-use unless the custom script/command changes `node_modules`, `package.json` or `package-lock.json`. For the static files build case for something like angular, it should generate assets to `<app>/dist` and this wouldn't be part of a layer and hence not cached.

Sounds good @arjun024 👍 . I think now that this would be in a separate buildpack, some of this stuff should be possible but we should certainly investigate more (I've only looked at these workflows with Angular in the past).

**Review by thitch97 — APPROVED** [review]

**Pull request merged**